src/persist/src/storage.rs

```diff
@@ -622,6 +664,104 @@ pub mod tests {
         Ok(())
     }
 
+    pub async fn consensus_impl_test<C: Consensus, F: FnMut() -> Result<C, Error>>(
+        mut new_fn: F,
+    ) -> Result<(), Error> {
+        let mut consensus = new_fn()?;
+
+        // Enforce that this entire test completes within 10 minutes.
+        let deadline = Instant::now() + Duration::from_secs(600);
+
+        // Starting value of consensus data is None.
+        assert_eq!(consensus.head(deadline).await, Ok(None));
+
+        let new_state = VersionedData {
+            seqno: SeqNo(5),
+            data: "abc".as_bytes().to_vec(),
+        };
+
+        // Incorrectly setting the data with a non-None expected should fail.
+        assert_eq!(
+            consensus
+                .compare_and_swap(Some(SeqNo(0)), Some(new_state.clone()), deadline)
+                .await,
+            Ok(Err(None))
+        );
+
+        // Correctly updating the state with the correct expected value should succeed.
+        assert_eq!(
+            consensus
+                .compare_and_swap(None, Some(new_state.clone()), deadline)
+                .await,
+            Ok(Ok(()))
+        );
+
+        // We can observe the a recent value on successful update.
+        assert_eq!(consensus.head(deadline).await, Ok(Some(new_state.clone())));
+
+        // It is allowed (though exceedingly strange) to set the sequence number back to
```

Comment:
I think this should be disallowed and we should document on the method that the new SeqNo must be strictly greater than the old one (new seqno < and = expected, with expected being correct, should then also be test cases)

Reply:
done!